### PR TITLE
Adds strength reduction and minimal nested counter rewrite rules

### PR DESCRIPTION
**Pull request by kavyasreedhar**
Nested counter rewrite is only tested for 2 nested counters so far. Some infrastructure to handle more than 2 counters is added so that the code can be extended for N nested counters and be more robust (will do in next PR)!

### CODE REVIEW

File: frail/ast.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from typing import Callable, Dict, Union
 
 
-@dataclass(eq=True, frozen=True)
```
Comment (David-Durst) [inline frail/ast.py:5]: Why did you need to unfreeze these? Are you modifying the nodes node-referential properties rather than creating new ones? Or is it just changing references?

Comment (kavyasreedhar) [reply frail/ast.py:5]: Yes, I'm not creating new nodes and am changing the nodes themselves. For example, in the strength rewrite, a node has arg0_index as a MulOp that is now been replaced by a counter val, this arg0_index is changed to be the counter index instead of the original multiply one.

Comment (David-Durst) [reply frail/ast.py:5]: That should work for the strength reduction rewrite because we're replacing equivalent nodes. However, for nested counter, I'm surprised you didn't run into issues. It's fine for now if no problems

In general, I prefer the dag to be immutable and then make a new dag after each rewrite. This makes debugging easier as you can look at the dag after each transformation and compare what happened. But not worth doing significant rewrites currently since your approach works.

File: frail/sharing_nested_counters_rewrite.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        cur_scan_lambda_var = var_f("scan_var_" + str(cur_scan_idx), e.width)
+        f_res = e.f(cur_scan_lambda_var)
+        e_ret = sharing_nested_counters_rewrite(f_res, False, lake_state)
+        e = scan_const_f(lambda z: e_ret)
+
+    elif e_type in (AddOp, SubOp, ModOp, EqOp, LTOp, GTOp, MulOp):
+        e.arg0_index = get_index(e.arg0_index, lake_state)
+        e.arg1_index = get_index(e.arg1_index, lake_state)
+        lake_state.program_map[e.index] = e
+        merged = get_merged_counter(lake_state)
```
Comment (David-Durst) [inline frail/sharing_nested_counters_rewrite.py:148]: please confirm my understanding: we're not modifying the addop/subop/mulop in this pass. We're only checking if their input is a merged counter, right? Because then we need to modify the input to these nodes?

File: frail/ast.py (hunk 1 above)
Comment (kavyasreedhar) [reply frail/ast.py:5]: Sounds good, can definitely add that as a todo for a future PR.

File: frail/sharing_nested_counters_rewrite.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+prev_vals = {}
+max_vals = {}
+replace = {}
+
+
+def nested_counters_rewrite(e: AST,
+                            root: bool = True,
+                            lake_state: LakeDSLState = default_lake_state):
+    need_merged_counters = True
+    while (need_merged_counters):
```
[the remark this answers is not in the document]
Comment (kavyasreedhar) [reply frail/sharing_nested_counters_rewrite.py:20]: Yes this executes exactly once. I was starting with doing N counters but then decided to work out issues for the 2 case first and make that solid and then just left this start here to help with the extension to N counters. I think accepting this PR and then extending this in a later pull makes sense.

File: frail/sharing_nested_counters_rewrite.py (hunk 2 above)
Comment (kavyasreedhar) [reply frail/sharing_nested_counters_rewrite.py:148]: Yes, exactly, that's right.

Comment (kavyasreedhar) [reply frail/sharing_nested_counters_rewrite.py:148]: A note though to follow up on that: the merged counter val actually replaces the add op when you are adding two counter vals that are now merged, so that is replaced here.

Comment (David-Durst) [reply frail/sharing_nested_counters_rewrite.py:148]: Makes sense. But why are you doing the same for mul and div? The rewrite rule only applies for add, right?

Comment (kavyasreedhar) [reply frail/sharing_nested_counters_rewrite.py:148]: That’s right, that’s a bug! Will update AddOp to be it’s own if statement with the merged code shortly.

Comment (David-Durst) [reply frail/sharing_nested_counters_rewrite.py:148]: Great! Approving. Please merge this once you've fixed that.

**Review by David-Durst — APPROVED** [review]

File: frail/sharing_nested_counters_rewrite.py (hunk 2 above)
Comment (kavyasreedhar) [reply frail/sharing_nested_counters_rewrite.py:148]: Pushed!
